pkg/cluster/update_master.go

```diff
@@ -0,0 +1,130 @@
+package cluster
+
+import (
+	"bytes"
+	"context"
+	"sort"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/cluster/kubeclient"
+	"github.com/openshift/openshift-azure/pkg/cluster/updateblob"
+	"github.com/openshift/openshift-azure/pkg/config"
+)
+
+func (u *simpleUpgrader) filterOldVMs(vms []compute.VirtualMachineScaleSetVM, blob *updateblob.UpdateBlob, ssHash []byte) []compute.VirtualMachineScaleSetVM {
+	var oldVMs []compute.VirtualMachineScaleSetVM
+	for _, vm := range vms {
+		if !bytes.Equal(blob.InstanceHashes[*vm.Name], ssHash) {
+			oldVMs = append(oldVMs, vm)
+		} else {
+			u.log.Infof("skipping vm %q since it's already updated", *vm.Name)
+		}
+	}
+	return oldVMs
+}
+
+// updateMasterAgentPool updates one by one all the VMs of the master scale set,
+// in place.
+func (u *simpleUpgrader) updateMasterAgentPool(ctx context.Context, cs *api.OpenShiftManagedCluster, app *api.AgentPoolProfile) *api.PluginError {
+	ssName := config.MasterScalesetName
+	ssHash, err := u.hasher.HashScaleSet(cs, app)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepUpdateMasterAgentPoolHashScaleSet}
+	}
+
+	blob, err := u.updateBlobService.Read()
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepUpdateMasterAgentPoolReadBlob}
+	}
+
+	vms, err := u.vmc.List(ctx, cs.Properties.AzProfile.ResourceGroup, ssName, "", "", "")
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepUpdateMasterAgentPoolListVMs}
+	}
+
+	// range our vms in order, so that if we previously crashed half-way through
+	// updating one and it is broken, we pick up where we left off.
+	sort.Slice(vms, func(i, j int) bool {
+		return *vms[i].VirtualMachineScaleSetVMProperties.OsProfile.ComputerName <
+			*vms[j].VirtualMachineScaleSetVMProperties.OsProfile.ComputerName
+	})
+
+	vms = u.filterOldVMs(vms, blob, ssHash)
```

Comment:
This part is a little bit confusing to me. If we hash each individual instances, we just pass one hash here and compare. Is the expectation that all instances hashes will match?

Reply:
ignore, tests explained